cpp/src/io/parquet/page_data.cu

```diff
@@ -1803,6 +1805,40 @@ cudaError_t PreprocessColumnData(hostdevice_vector<PageInfo> &pages,
   // back, this value will get overwritten later on).
   pages.device_to_host(stream, true);
 
+  // ordering of pages is by input column schema, repeated across row groups.  so
+  // if we had 3 columns, each with 2 pages, and 1 row group, our schema values might look like
+  //
+  // 1, 1, 2, 2, 3, 3
+  //
+  // However, if we had more than one row group, the pattern would be
+  //
+  // 1, 1, 2, 2, 3, 3, 1, 1, 2, 2, 3, 3
+  // ^ row group 0     |
+  //                   ^ row group 1
+  //
+  // To use exclusive_scan_by_key, the ordering we actually want is
+  //
+  // 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3
+  //
+  // We also need to preserve key-relative page ordering, so we need to use a stable sort.
+  thrust::device_vector<int> page_keys(pages.size());
```

Comment:
Since these don't appear to need initialization, please use `rmm::device_uvector` since it saves synchronization and initialization, and takes an explicit stream.

Reply:
BTW, you should always use `rmm::device_vector` instead of `thrust::device_vector` in RAPIDS code anyway.  It's a simple wrapper that uses the RMM allocator for the vector.  In this case `uvector` is preferred though.